File: rust/sort/bubble_sort.rs
/// 冒泡排序 (Bubble Sort)
/// 
/// # 算法原理
/// 冒泡排序是一种简单的排序算法。它重复地遍历要排序的列表，
/// 比较相邻的两个元素，如果它们的顺序错误就把它们交换过来。
/// 遍历列表的工作是重复进行的，直到没有再需要交换的元素，
/// 这意味着列表已经排序完成。
/// 
/// # 时间复杂度
/// - 最好情况: O(n) - 当数组已经有序时
/// - 最坏情况: O(n²) - 当数组逆序时
/// - 平均情况: O(n²)
/// 
/// # 空间复杂度
/// O(1) - 只需要常数级别的额外空间
/// 
/// # 稳定性
/// 稳定排序 - 相等元素的相对顺序不会改变

/// 冒泡排序函数
/// 
/// # 参数
/// - `arr`: 待排序的可变数组切片
/// 
/// # 示例
/// ```
/// let mut arr = vec![64, 34, 25, 12, 22, 11, 90];
/// bubble_sort(&mut arr);
/// assert_eq!(arr, vec![11, 12, 22, 25, 34, 64, 90]);
/// ```
fn bubble_sort<T: Ord>(arr: &mut [T]) {
    let n = arr.len();
    
    // 如果数组长度小于等于1，无需排序
    if n <= 1 {
        return;
    }
    
    // 外层循环：控制排序的轮数
    // 每一轮都会把当前未排序部分的最大值"冒泡"到末尾
    for i in 0..n {
        // 优化标志：如果某一轮没有发生交换，说明数组已经有序
        let mut swapped = false;
        
        // 内层循环：比较相邻元素并交换
        // 注意：n - 1 - i 是因为每轮排序后，末尾的 i 个元素已经有序
        for j in 0..n - 1 - i {
            // 如果前一个元素大于后一个元素，则交换它们
            if arr[j] > arr[j + 1] {
                arr.swap(j, j + 1);
                swapped = true;
            }
        }
        
        // 如果这一轮没有发生任何交换，说明数组已经有序，提前退出
        if !swapped {
            break;
        }
    }
}

/// 主函数 - 演示冒泡排序的使用
fn main() {
    println!("=== 冒泡排序 (Bubble Sort) 演示 ===\n");
    
    // 示例1：整数数组排序
    let mut numbers = vec![64, 34, 25, 12, 22, 11, 90];
    println!("原始数组: {:?}", numbers);
    bubble_sort(&mut numbers);
    println!("排序后:   {:?}", numbers);
    
    println!();
    
    // 示例2：已经有序的数组（最好情况）
    let mut sorted = vec![1, 2, 3, 4, 5];
    println!("已有序数组: {:?}", sorted);
    bubble_sort(&mut sorted);
    println!("排序后:     {:?}", sorted);
    
    println!();
    
    // 示例3：逆序数组（最坏情况）
    let mut reversed = vec![5, 4, 3, 2, 1];
    println!("逆序数组: {:?}", reversed);
    bubble_sort(&mut reversed);
    println!("排序后:   {:?}", reversed);
    
    println!();
    
    // 示例4：字符串数组排序
    let mut fruits = vec!["banana", "apple", "cherry", "date"];
    println!("字符串数组: {:?}", fruits);
    bubble_sort(&mut fruits);
    println!("排序后:     {:?}", fruits);
    
    println!();
    
    // 示例5：空数组和单元素数组
    let mut empty: Vec<i32> = vec![];
    let mut single = vec![42];
    println!("空数组: {:?}", empty);
    bubble_sort(&mut empty);
    println!("排序后: {:?}", empty);
    println!("单元素数组: {:?}", single);
    bubble_sort(&mut single);
    println!("排序后:     {:?}", single);
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_normal_array() {
        let mut arr = vec![64, 34, 25, 12, 22, 11, 90];
        bubble_sort(&mut arr);
        assert_eq!(arr, vec![11, 12, 22, 25, 34, 64, 90]);
    }
    
    #[test]
    fn test_empty_array() {
        let mut arr: Vec<i32> = vec![];
        bubble_sort(&mut arr);
        assert_eq!(arr, vec![]);
    }
    
    #[test]
    fn test_single_element() {
        let mut arr = vec![42];
        bubble_sort(&mut arr);
        assert_eq!(arr, vec![42]);
    }
    
    #[test]
    fn test_sorted_array() {
        let mut arr = vec![1, 2, 3, 4, 5];
        bubble_sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 3, 4, 5]);
    }
    
    #[test]
    fn test_reversed_array() {
        let mut arr = vec![5, 4, 3, 2, 1];
        bubble_sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 3, 4, 5]);
    }
    
    #[test]
    fn test_with_duplicates() {
        let mut arr = vec![3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5];
        bubble_sort(&mut arr);
        assert_eq!(arr, vec![1, 1, 2, 3, 3, 4, 5, 5, 5, 6, 9]);
    }
}

File: cpp/bubble_sort.cpp
/**
 * 冒泡排序 (Bubble Sort)
 * 
 * 算法原理：
 * 冒泡排序是一种简单的排序算法。它重复地遍历要排序的数列，一次比较两个元素，
 * 如果它们的顺序错误就把它们交换过来。遍历数列的工作是重复进行的，直到没有
 * 再需要交换的元素为止，说明数列已经排序完成。
 * 
 * 算法步骤：
 * 1. 比较相邻的元素。如果第一个比第二个大，就交换它们。
 * 2. 对每一对相邻元素做同样的工作，从开始第一对到结尾的最后一对。
 *    这步做完后，最后的元素会是最大的数。
 * 3. 针对所有的元素重复以上的步骤，除了最后一个。
 * 4. 持续每次对越来越少的元素重复上面的步骤，直到没有任何一对数字需要比较。
 * 
 * 时间复杂度：
 * - 最好情况：O(n) - 当数组已经有序时
 * - 最坏情况：O(n²) - 当数组逆序时
 * - 平均情况：O(n²)
 * 
 * 空间复杂度：O(1) - 只需要常数级别的额外空间
 * 
 * 稳定性：稳定 - 相等元素的相对顺序不会改变
 */

#include <iostream>
#include <vector>

/**
 * 冒泡排序函数
 * @param arr 待排序的数组（引用传递，直接修改原数组）
 */
void bubbleSort(std::vector<int>& arr) {
    int n = arr.size();
    
    // 外层循环：控制排序的轮数
    // 每完成一轮，就有一个最大元素被"冒泡"到正确位置
    for (int i = 0; i < n - 1; i++) {
        // 优化：设置一个标志位，如果某一轮没有发生交换，说明已经有序
        bool swapped = false;
        
        // 内层循环：进行相邻元素的比较和交换
        // 注意：每轮排序后，最后i个元素已经有序，无需再比较
        for (int j = 0; j < n - 1 - i; j++) {
            // 如果前一个元素大于后一个元素，则交换它们
            if (arr[j] > arr[j + 1]) {
                std::swap(arr[j], arr[j + 1]);
                swapped = true;  // 标记发生了交换
            }
        }
        
        // 如果这一轮没有发生任何交换，说明数组已经有序，提前退出
        if (!swapped) {
            break;
        }
    }
}

/**
 * 打印数组的辅助函数
 * @param arr 要打印的数组
 * @param message 打印前的提示信息
 */
void printArray(const std::vector<int>& arr, const std::string& message) {
    std::cout << message;
    for (int num : arr) {
        std::cout << num << " ";
    }
    std::cout << std::endl;
}

/**
 * 主函数 - 演示冒泡排序
 */
int main() {
    // 创建测试数组
    std::vector<int> arr = {64, 34, 25, 12, 22, 11, 90, 5, 77, 30};
    
    std::cout << "========== 冒泡排序演示 ==========" << std::endl;
    printArray(arr, "排序前: ");
    
    // 执行冒泡排序
    bubbleSort(arr);
    
    printArray(arr, "排序后: ");
    
    // 测试已排序数组（展示优化效果）
    std::cout << "\n测试已排序数组：" << std::endl;
    std::vector<int> sorted = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    printArray(sorted, "排序前: ");
    bubbleSort(sorted);
    printArray(sorted, "排序后: ");
    
    // 测试逆序数组
    std::cout << "\n测试逆序数组：" << std::endl;
    std::vector<int> reversed = {10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
    printArray(reversed, "排序前: ");
    bubbleSort(reversed);
    printArray(reversed, "排序后: ");
    
    return 0;
}
